modules/core/src/main/java/org/locationtech/jts/coverage/CoverageSimplifier.java

```diff
@@ -58,93 +72,209 @@ public class CoverageSimplifier {
+    this.smoothWeight  = smoothWeight;
+  }
+  
+  /**
+   * Computes the simplified coverage using a single distance tolerance, 
+   * preserving the coverage topology.
+   * 
+   * @param tolerance the simplification distance tolerance
+   * @return the simplified coverage polygons
+   */
+  public Geometry[] simplify(double tolerance) {
+    return simplifyEdges(tolerance, tolerance);
+  }
 
-    simplifyEdges(innerEdges, constraintEdges, tolerance);
-    Geometry[] result = cov.buildCoverage();
-    return result;
+  /**
+   * Computes the simplified coverage using separate distance tolerances
+   * for inner and outer edges, 
+   * preserving the coverage topology.
+   * 
+   * @param toleranceInner the distance tolerance for inner edges
+   * @param toleranceOuter the distance tolerance for outer edges
+   * @return the simplified coverage polygons
+   */
+  public Geometry[] simplify(double toleranceInner, double toleranceOuter) {
+    return simplifyEdges(toleranceInner, toleranceOuter);
+  }
+
+  /**
+   * Computes the simplified coverage using separate distance tolerances
+   * for each coverage element, 
+   * preserving the coverage topology.
+   * 
+   * @param tolerances the distance tolerances for the coverage elements
+   * @return the simplified coverage polygons
+   */
+  public Geometry[] simplify(double[] tolerances) {
+    if (tolerances.length != coverage.length)
+      throw new IllegalArgumentException("number of tolerances does not match number of coverage elements");
+    return simplifyEdges(tolerances);
+  }
+
+  private Geometry[] simplifyEdges(double[] tolerances) {
+    CoverageRingEdges covRings = CoverageRingEdges.create(coverage);
+    List<CoverageEdge> covEdges = covRings.getEdges();
+    TPVWSimplifier.Edge[] edges = createEdges(covEdges, tolerances);
+    return simplify(covRings, covEdges, edges);
+  }
+
+  private Edge[] createEdges(List<CoverageEdge> covEdges, double[] tolerances) {
+    TPVWSimplifier.Edge[] edges = new TPVWSimplifier.Edge[covEdges.size()];
+    for (int i = 0; i < covEdges.size(); i++) {
+      CoverageEdge covEdge = covEdges.get(i);
+      double tol = computeTolerance(covEdge, tolerances);
+      edges[i] = createEdge(covEdge, tol);
+    }
+    return edges;
   }
 
-  private void simplifyEdges(List<CoverageEdge> edges, MultiLineString constraints, double tolerance) {
-    MultiLineString lines = CoverageEdge.createLines(edges, geomFactory);
-    BitSet freeRings = getFreeRings(edges);
-    MultiLineString linesSimp = TPVWSimplifier.simplify(lines, freeRings, constraints, tolerance);
-    //Assert: mlsSimp.getNumGeometries = edges.length
+  private double computeTolerance(CoverageEdge covEdge, double[] tolerances) {
+    int index0 = covEdge.getAdjacentIndex(0);
+    // assert: index0 >= 0
+    double tolerance = tolerances[index0];
     
-    setCoordinates(edges, linesSimp);
+    int index1 = covEdge.getAdjacentIndex(0);
```

Comment:
int index1 = covEdge.getAdjacentIndex(1);

Reply:
Good catch.  Need a unit test to verify that.